File: Windows/universal8.1/Siminov/Core/Database/DatabaseUtils.cs
/** 
 * [SIMINOV FRAMEWORK]
 * Copyright [2015] [Siminov Software Solution LLP|support@siminov.com]
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 **/


#if __MOBILE__
#define XAMARIN
#endif

#if !__MOBILE__
#define WINDOWS
#endif


using Siminov.Core.Utils;
using Siminov.Core.Model;
using Siminov.Core.Resource;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;


#if XAMARIN
using PCLStorage;
#endif

namespace Siminov.Core.Database
{

    /// <summary>
    /// Exposes utils methods to deal with database. 
    /// It has methods to form database file paths.
    /// </summary>
    public class DatabaseUtils
    {

        /// <summary>
        /// Check whether SDCard is present on device or not.
        /// </summary>
        /// <returns>TRUE: If SDCard is present on device, FALSE: If SDCard is not present.</returns>
        public bool IsSDCardPresent()
        {
            return false;//android.os.Environment.getExternalStorageState().equals(android.os.Environment.MEDIA_MOUNTED);
        }

        /// <summary>
        /// Get internal memory database path
        /// </summary>
        /// <param name="databaseDescriptor">Database Descriptor Object</param>
        /// <returns></returns>
        public String InternalMemoryDatabasePath(DatabaseDescriptor databaseDescriptor)
        {

            #if XAMARIN
            
                var rootFolder = FileSystem.Current.LocalStorage;
			    return rootFolder.Path;

            #elif WINDOWS

                ResourceManager resourceManager = ResourceManager.GetInstance();
                ApplicationDescriptor applicationDescriptor = resourceManager.GetApplicationDescriptor();

                String databaseDirName = databaseDescriptor.GetDatabaseName();

                String databaseDirPath = applicationDescriptor.GetName() + FileUtils.Separator + Constants.DATABASE_PATH_DATABASE + FileUtils.Separator + databaseDirName + FileUtils.Separator;
                return databaseDirPath;
            
            #endif

        }

        /// <summary>
        /// Get sdcard database path
        /// </summary>
        /// <param name="databaseDescriptor">Database Descriptor Object</param>
        /// <returns></returns>
        public String SdcardMemoryDatabasePath(DatabaseDescriptor databaseDescriptor)
        {
            /*File externalStorage = Environment.GetExternalStorageDirectory();
            String externalStoragePath = externalStorage.GetAbsolutePath();

            ResourceManager resourceManager = ResourceManager.GetInstance();
            ApplicationDescriptor applicationDescriptor = resourceManager.GetApplicationDescriptor();

            String databaseDirName = databaseDescriptor.GetDatabaseName();

            String databaseDirPath = externalStoragePath + File.separator + applicationDescriptor.getName() + File.separator + Constants.DATABASE_PATH_DATABASE + File.separator + databaseDirName + File.separator;
            return databaseDirPath;*/
            return "";
        }

        /// <summary>
        /// Get database path based on which storage needed
        /// </summary>
        /// <param name="databaseDescriptor">Database Descriptor Object</param>
        /// <returns></returns>
        public String GetDatabasePath(DatabaseDescriptor databaseDescriptor)
        {
            DatabaseUtils databaseUtils = new DatabaseUtils();
            String databasePath = null;

            bool isExternalStorageEnable = databaseDescriptor.IsExternalStorageEnable();
            if (isExternalStorageEnable)
            {
                if (databaseUtils.IsSDCardPresent())
                {
                    databasePath = databaseUtils.SdcardMemoryDatabasePath(databaseDescriptor);
                }
                else
                {
                    databasePath = databaseUtils.InternalMemoryDatabasePath(databaseDescriptor);
                }
            }
            else
            {
                databasePath = databaseUtils.InternalMemoryDatabasePath(databaseDescriptor);
            }

            return databasePath;
        }

    }
}

File: Android/Core/src/main/java/siminov/core/database/DatabaseUtils.java
/**
 * [SIMINOV FRAMEWORK - CORE]
 * Copyright [2014-2016] [Siminov Software Solution LLP|support@siminov.com]
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 **/


package siminov.core.database;

import java.io.File;

import siminov.core.Constants;
import siminov.core.model.ApplicationDescriptor;
import siminov.core.model.DatabaseDescriptor;
import siminov.core.resource.ResourceManager;
import android.content.Context;
import android.os.Environment;


/**
 * Exposes utils methods to deal with database. 
 * It has methods to form database file paths.
 */
public class DatabaseUtils {

	/**
	 * Check whether SDCard is present on device or not.
	 * @return TRUE: If SDCard is present on device, FALSE: If SDCard is not present.
	 */
	public boolean isSDCardPresent() {
		return android.os.Environment.getExternalStorageState().equals(android.os.Environment.MEDIA_MOUNTED);
	}
	
	/**
	 * Get internal memory database path.
	 * @param databaseDescriptor Database Descriptor Object.
	 */
	public String internalMemoryDatabasePath(final DatabaseDescriptor databaseDescriptor) {
		ResourceManager resourceManager = ResourceManager.getInstance();
		ApplicationDescriptor applicationDescriptor = resourceManager.getApplicationDescriptor();
		Context context = resourceManager.getApplicationContext();
		
		String databaseDirName = databaseDescriptor.getDatabaseName();
		
		String databaseDirPath = Constants.DATABASE_PATH_DATA + File.separator + Constants.DATABASE_PATH_DATA + File.separator + context.getPackageName() + File.separator + applicationDescriptor.getName() + File.separator + Constants.DATABASE_PATH_DATABASE + File.separator + databaseDirName + File.separator;
		return databaseDirPath;
	}
	
	/**
	 * Get sdcard database path.
	 * @param databaseDescriptor Database Descriptor Object.
	 */
	public String sdcardMemoryDatabasePath(final DatabaseDescriptor databaseDescriptor) {
		File externalStorage = Environment.getExternalStorageDirectory();
		String externalStoragePath = externalStorage.getAbsolutePath();
		
		ResourceManager resourceManager = ResourceManager.getInstance();
		ApplicationDescriptor applicationDescriptor = resourceManager.getApplicationDescriptor();

		String databaseDirName = databaseDescriptor.getDatabaseName();

		String databaseDirPath = externalStoragePath + File.separator + applicationDescriptor.getName() + File.separator + Constants.DATABASE_PATH_DATABASE + File.separator + databaseDirName + File.separator;
		return databaseDirPath;
	}
	
	/**
	 * Get database path based on which storage needed.
	 * @param databaseDescriptor Database Descriptor Object.
	 */
	public String getDatabasePath(final DatabaseDescriptor databaseDescriptor) {
		DatabaseUtils databaseUtils = new DatabaseUtils();
		String databasePath = null;

		boolean isExternalStorageEnable = databaseDescriptor.isExternalStorageEnable();
		if(isExternalStorageEnable) {
			if(databaseUtils.isSDCardPresent()) {
				databasePath = databaseUtils.sdcardMemoryDatabasePath(databaseDescriptor);
			} else {
				databasePath = databaseUtils.internalMemoryDatabasePath(databaseDescriptor);
			}
		} else {
			databasePath = databaseUtils.internalMemoryDatabasePath(databaseDescriptor);
		}

		return databasePath;
	}
}
